### PR TITLE
Calculate visibilities once in resolve

### PR DESCRIPTION
**Pull request by petrochenkov**
Then use them through a query based on resolver outputs.

Item visibilities were previously calculated in three places - initially in `rustc_resolve`, then in `rustc_privacy` during type privacy checkin, and then in `rustc_metadata` during metadata encoding.
The visibility logic is not entirely trivial, especially for things like constructors or enum variants, and all of it was duplicated.

This PR deduplicates all the visibility calculations, visibilities are determined once during early name resolution and then stored in `ResolverOutputs` and are later available through `tcx` as a query `tcx.visibility(def_id)`.
(This query existed previously, but only worked for other crates.)

Some special cases (e.g. visibilities for closure types, which are needed for type privacy checking) are not processed in resolve, but deferred and performed directly in the query instead.

### CODE REVIEW

File: compiler/rustc_privacy/src/lib.rs [hunk 1]
```diff
@@ -2064,13 +1957,61 @@ impl<'a, 'tcx> Visitor<'tcx> for PrivateItemsInPublicInterfacesVisitor<'a, 'tcx>
...
+                Node::ImplItem(impl_item) => {
+                    match tcx.hir().get(tcx.hir().get_parent_item(hir_id)) {
+                        Node::Item(hir::Item {
+                            kind: hir::ItemKind::Impl { of_trait: Some(trait_ref), .. },
+                            ..
+                        }) => trait_ref
+                            .path
+                            .res
+                            .opt_def_id()
+                            .map_or(ty::Visibility::Public, |def_id| tcx.visibility(def_id)),
```
Comment (lcnr) [inline compiler/rustc_privacy/src/lib.rs:2001]: is `opt_def_id` only `None` here if we already emitted an error?

Can we use a delay_span_bug here?

**Review by lcnr — COMMENTED** [review]
So looking through this I didn't see anything which looked out of place, but I am not familiar enough with this code to actually review these changes :thinking:

**Comment by lcnr** [comment]
https://github.com/rust-lang/compiler-team/blob/master/content/experts/map.toml mentions @davidtwco but feel free to reassign

r? @davidtwco for now

**Comment by petrochenkov** [comment]
Question: is it possible to make a query `eval_always` for the current crate and not `eval_always` for other crates?
(Or even make it use some more complex condition?)

As I understand, the query must be `eval_always` because it accesses data from `ResolverOutputs` which is built before query system starts working, but for visibilities from other crates it would be undesirable to decode them from metadata again and again.

File: src/test/ui/pub/issue-33174-restricted-type-in-public-interface.rs [hunk 2]
```diff
@@ -16,11 +16,11 @@ pub struct Shell<T> {
...
 pub type Helix_pomatia = Shell<Snail>;
-//~^ ERROR crate-visible type `Snail` in public interface
-//~| NOTE can't leak crate-visible type
+//~^ ERROR private type `Snail` in public interface
+//~| NOTE can't leak private type
 pub type Dermochelys_coriacea = Shell<sea::Turtle>;
-//~^ ERROR restricted type `Turtle` in public interface
-//~| NOTE can't leak restricted type
+//~^ ERROR crate-private type `Turtle` in public interface
+//~| NOTE can't leak crate-private type
```
Comment (jyn514) [inline src/test/ui/pub/issue-33174-restricted-type-in-public-interface.rs:23]: Are these changes intentional?

File: src/test/ui/proc-macro/issue-50493.stderr [hunk 3]
```diff
@@ -1,26 +1,9 @@
...
-   |
-   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
-
-error[E0616]: field `field` of struct `Restricted` is private
-  --> $DIR/issue-50493.rs:6:10
-   |
-LL | #[derive(Derive)]
-   |          ^^^^^^ private field
-   |
-   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```
Comment (jyn514) [inline src/test/ui/proc-macro/issue-50493.stderr:21]: Is this change in the diagnostic intentional?

File: src/test/ui/pub/issue-33174-restricted-type-in-public-interface.rs (hunk 2 above)
Comment (petrochenkov) [reply src/test/ui/pub/issue-33174-restricted-type-in-public-interface.rs:23]: The wording is now based on semantic visibilities returned by `tcx.visibility()`.
Previously it was based on syntactic visibilities.
You can see that the visibility in question is actually crate-private (`pub(crate)`), even if it's written differently syntactically.

File: src/test/ui/proc-macro/issue-50493.stderr (hunk 3 above)
Comment (petrochenkov) [reply src/test/ui/proc-macro/issue-50493.stderr:21]: Visibility with incorrect path `restricted` is now recovered as public, which prevents the follow-up errors.

File: compiler/rustc_privacy/src/lib.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_privacy/src/lib.rs]: Done.

**Review by davidtwco — APPROVED** [review]
LGTM, r=me if you're happy with this.

**Comment by Mark-Simulacrum** [comment]
Final perf [results](https://perf.rust-lang.org/compare.html?start=1d2726726f8f3128e98191e4c6cb94bd76d0ddd4&end=1eaadebb3dee31669c7649b32747381d11614fae&stat=instructions:u) were approximately equivalent to the perf run on this PR; some solid improvement on deeply-nested-async benchmark in particular. Good work!
